File: myfinance-mobile/test/core/relative_age_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:myfinance/core/format/money.dart';

/// The table both implementations are pinned to.
///
/// The widget computes the age in Kotlin, because a string formatted at sync
/// time would freeze on "just now". Two implementations means two chances to
/// drift, so both are held to this same list — see `RelativeAgeTest.kt`.
const cases = <(Duration, String)>[
  (Duration(seconds: 0), 'just now'),
  (Duration(seconds: 59), 'just now'),
  (Duration(minutes: 1), '1 min ago'),
  (Duration(minutes: 59), '59 min ago'),
  (Duration(hours: 1), '1 hour ago'),
  (Duration(hours: 2), '2 hours ago'),
  (Duration(hours: 23), '23 hours ago'),
  (Duration(days: 1), '1 day ago'),
  (Duration(days: 29), '29 days ago'),
  (Duration(days: 30), 'over a month ago'),
  (Duration(days: 400), 'over a month ago'),
];

void main() {
  final now = DateTime(2026, 8, 20, 12);

  group('formatRelativeAge', () {
    for (final (elapsed, expected) in cases) {
      test('$elapsed reads as "$expected"', () {
        expect(formatRelativeAge(now.subtract(elapsed), now: now), expected);
      });
    }

    test('a timestamp in the future is not a negative age', () {
      // A snapshot written a moment ahead of this device's clock, or a clock
      // that moved backwards. Neither is worth "-3 minutes ago".
      expect(
        formatRelativeAge(now.add(const Duration(minutes: 5)), now: now),
        'just now',
      );
    });

    test('the boundaries fall on the right side', () {
      expect(formatRelativeAge(now.subtract(const Duration(minutes: 60)), now: now),
          '1 hour ago');
      expect(formatRelativeAge(now.subtract(const Duration(hours: 24)), now: now),
          '1 day ago');
    });
  });
}

File: myfinance-mobile/android/app/src/test/kotlin/com/methou/myfinance/widget/RelativeAgeTest.kt
package com.methou.myfinance.widget

import java.time.Duration
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

/**
 * The same table `relative_age_test.dart` holds the Dart side to.
 *
 * The widget has to work the age out itself — a string formatted at sync time
 * would freeze on "just now" — so the wording exists twice. Keeping both pinned
 * to identical cases means neither can drift without one of these two files
 * going red.
 */
class RelativeAgeTest {

    private val cases = listOf(
        Duration.ofSeconds(0) to "just now",
        Duration.ofSeconds(59) to "just now",
        Duration.ofMinutes(1) to "1 min ago",
        Duration.ofMinutes(59) to "59 min ago",
        Duration.ofHours(1) to "1 hour ago",
        Duration.ofHours(2) to "2 hours ago",
        Duration.ofHours(23) to "23 hours ago",
        Duration.ofDays(1) to "1 day ago",
        Duration.ofDays(29) to "29 days ago",
        Duration.ofDays(30) to "over a month ago",
        Duration.ofDays(400) to "over a month ago",
    )

    @Test
    fun `matches the wording the app uses`() {
        for ((elapsed, expected) in cases) {
            assertEquals("for $elapsed", expected, RelativeAge.describe(elapsed))
        }
    }

    @Test
    fun `a timestamp in the future is not a negative age`() {
        assertEquals("just now", RelativeAge.describe(Duration.ofMinutes(-5)))
    }

    @Test
    fun `an unreadable timestamp yields nothing rather than a lie`() {
        // The header then shows no age at all, which is honest. Printing
        // "just now" for a timestamp we could not read would not be.
        assertNull(RelativeAge.format(null))
        assertNull(RelativeAge.format(""))
        assertNull(RelativeAge.format("not a timestamp"))
    }

    @Test
    fun `reads the ISO local timestamp Dart writes`() {
        val iso = "2026-08-20T09:00:00.000"
        assertEquals(
            "the format Dart's toIso8601String produces must parse",
            true,
            RelativeAge.format(iso) != null,
        )
    }
}
